### ISSUE TITLE
Suggested Changes 1/11/21

### ISSUE THREAD

**Issue by crockeea** [cfrg/draft-irtf-cfrg-opaque#112]
I read through the draft (https://cfrg.github.io/draft-irtf-cfrg-opaque/draft-irtf-cfrg-opaque.html on 1/11/21), and have a few comments.

-	Add `pwdU` to section 1.2
-	Add OPRF `KeyGen()` to section 2 (referenced in section 3.2.2.2)
-	The type of `mode` in the `InnerEnvelope` struct should be `EnvelopeMode`
-	Section 3.1: The text above the definition of the `Envelope` and `InnerEnvelope` structs says, “These credentials are embedded into the following structure…”. I therefore expected to see a `SecretCredentials` struct and `CleartextCredentials` struct in either `Envelope` or `InnerEnvelope`. There is a field for “encrypted_creds”, which could correspond to the encryption of the `SecretCredentials`, but there still seems to be no place where `CleartextCredentials` can be embedded into these structs.
-	Be consistent about the use of parens for functions which output multiple values (see protocol diagrams at top of section 3.2 and at the top of 3.3)
-	Notation for kU: section 1.2 says “kU refers to U's private OPRF key”. Without any other context, I expect this to be something generated by the user, and known only to the user. But in reality, kU is a per-user key generated by the server, and not revealed to the client. Is there a different way to describe `kU` to avoid this potential confusion?
-	OPRF blind is referred to as both `blind` and `r`. Make this consistent for clarity.
-	Need to define the `Credentials` structure mentioned at the end of section 3.1: fields of this struct are referenced in section 3.2.2.3, but never defined.
-	Section 3.2.2.3 step 7: uses `len(pt)`, but I’m not sure what `pt` is supposed to be/refer to. Perhaps `secret_creds`?
-	Section 3.2.2.3 step 11: “Create InnerEnvelope contents”. I read this as “Create [an unnamed] InnerEnvelope body….”, but you apparently intended to name the struct `contents`, since you then refer to it in step 12. If you gave this a non-English-word name, this would be clearer, since there are no typesetting clues.
-	Is it really necessary to define a variant of `CleartextCredentials`? Why not always use the larger struct? This is only included as AAD; it is not actually transmitted or stored. It’s adding a lot of extra complexity with no apparent benefit. Implementations can always used a short fixed value if they would otherwise want to use the truncated `CleartextCredentials`.
-	The type of OPRF group elements isn’t consistent: in section 3.2.1, it’s an (undefined!) type `SerializedElement data`; in section 3.3.1, it’s `opaque data<1..2^16-1>`
-	Section 3.3.2.3: The “Outputs” list a `SecretCredentials` and an export key, but the diagram shows that RecoverCredentials outputs (skU, pkS, export_key)
-	Suggestion: incorporate generic AKE messages into the protocol in section 3.3. It would be nice to have a complete diagram of the generic protocol. Currently, you have a partial diagram in section 3.3 and a textual description of the full protocol in the intro of section 4.
-	Section 4 intro: “However, an instantiation where both AKE and data encryption are quantum safe…” Is this encryption referring to the EnvU encryption defined in section 3.2.2.3? If so, isn’t the encryption already quantum safe? If that’s the case, I’d change this to “However, an instantiation where AKE is quantum safe…” to avoid confusion.
-	Section 4.1: struct `HkdfLabel` contains the field opaque label<8..255> = "OPAQUE " + Label; Is `Label` required to be non-null, or is the string required to be null-terminated? The string “OPAQUE “ has length 7, so I’m wondering why the minimum field size is 8. (I see that this is consistent with TLS1.3, but I don’t see an explanation there either.)
-	Section 4.2.1: This section defines, e.g., ` KE1 = credential_request, nonceU, info1*, epkU`, so that KE1 contains the credential_request. However, at the top of section 4, the high-level protocol shows credential_request being sent in addition to KE1. This is another reason it would be helpful to have a complete diagram of the generic protocol (with, e.g., a generic 3-message AKE). For this particular issue, I think KE1 should only include AKE message components; it should not include the credential_request. Similarly for KE2 and section 4.3.1.
-	Section 5: “Future specifications … MAY introduce other AKE instantiations.” Also allow for/mention the possibility of defining alternate OPRFs. For example, https://eprint.iacr.org/2020/1532 gives a lattice-based (and hence PQ) OPRF.
-	Section 6.6: What does “skipping the key exchange part” mean? Does it mean somehow truncating the protocol/skipping the AKE entirely? Or does it mean not using the exchanged key after a successful authentication? In general, I’m not sure what I would have to do to violate this section, which means I could easily violate it without knowing.

**Comment by chris-wood** [comment]
Thanks, @crockeea, for the detailed review! I reviewed the comments and turned those applicable into action items below. (Some are already addressed, so I omitted those.) I responded to specific comments below that list. Please let me know if I misunderstood any of your feedback.

Tasks:
- [ ] Add pwdU to section 1.2 (notation)
- [ ] The type of mode in the InnerEnvelope struct should be EnvelopeMode
- [ ] Section 3.1: Clarify that not all credentials are embedded in the envelope
- [ ] Be consistent about the use of parens for functions which output multiple values (see protocol diagrams at top of section 3.2 and at the top of 3.3)
- [ ] Clarify kU ownership
- [ ] Section 3.2.2.3 step 7: s/len(pt)/len(secret_creds)
- [ ] Section 3.2.2.3 step 11: Rename InnerEnvelope from "contents" to something less confusing, perhaps "inner_envelope"?
- [ ] Update Section 4 OPRF+AKE text description with a diagram like that in the registration and authentication steps.
- [ ] Clarify HkdfLabel "Label" properties (Is Label required to be non-null, or is the string required to be null-terminated?)
- [ ] Section 5: Allow future OPRFs to be chosen, possibly citing PQ OPRFs (https://eprint.iacr.org/2020/1532).

Comments:

> Add OPRF KeyGen() to section 2 (referenced in section 3.2.2.2)

This is already in Section 2, where we say, "We also assume the existence of a function KeyGen from [I-D.irtf-cfrg-voprf], which generates an OPRF private and public key."

> Section 3.1: The text above the definition of the Envelope and InnerEnvelope structs says, “These credentials are embedded into the following structure…”. I therefore expected to see a SecretCredentials struct and CleartextCredentials struct in either Envelope or InnerEnvelope. There is a field for “encrypted_creds”, which could correspond to the encryption of the SecretCredentials, but there still seems to be no place where CleartextCredentials can be embedded into these structs.

Good point! Only the secret credentials are actually included in the envelope. The cleartext credentials are omitted but ultimately folded into the MAC check. We can clarify the text to make this more clear.

> Notation for kU: section 1.2 says “kU refers to U's private OPRF key”. Without any other context, I expect this to be something generated by the user, and known only to the user. But in reality, kU is a per-user key generated by the server, and not revealed to the client. Is there a different way to describe kU to avoid this potential confusion?

Great point. We can probably say something along the lines of, "kU is a unique OPRF key for user U known only to the server."

> Need to define the Credentials structure mentioned at the end of section 3.1: fields of this struct are referenced in section 3.2.2.3, but never defined.

Credentials is meant to be an implementation-specific construct with fields listed in its description. We chose to not define this structure using standard notation since it's irrelevant for the description protocol. All that matters is that the core protocol can access the name fields from a Credentials object. (We do [something similar for HPKE] when generating encryption contexts.)

> Is it really necessary to define a variant of CleartextCredentials? Why not always use the larger struct? This is only included as AAD; it is not actually transmitted or stored. It’s adding a lot of extra complexity with no apparent benefit. Implementations can always used a short fixed value if they would otherwise want to use the truncated CleartextCredentials.

Hmm, I'm not sure I understand this comment. @crockeea, would you mind elaborating?

> Suggestion: incorporate generic AKE messages into the protocol in section 3.3. It would be nice to have a complete diagram of the generic protocol. Currently, you have a partial diagram in section 3.3 and a textual description of the full protocol in the intro of section 4.

I think it's more readable if the core protocol description is unencumbered by the AKE steps. How about a compromise: replace the textual description in Section 4 with a graphical description similar to that in Section 3?

> Section 4 intro: “However, an instantiation where both AKE and data encryption are quantum safe…” Is this encryption referring to the EnvU encryption defined in section 3.2.2.3? If so, isn’t the encryption already quantum safe? If that’s the case, I’d change this to “However, an instantiation where AKE is quantum safe…” to avoid confusion.

I think this is referring to the encryption of data sent during and after the AKE, though I'm now not certain :) @hugokraw, can you double check this paragraph?

> Section 4.1: struct HkdfLabel contains the field opaque label<8..255> = "OPAQUE " + Label; Is Label required to be non-null, or is the string required to be null-terminated? The string “OPAQUE “ has length 7, so I’m wondering why the minimum field size is 8. (I see that this is consistent with TLS1.3, but I don’t see an explanation there either.)

Hmm, good question. I don't recall off hand. I filed [this issue](https://github.com/tlswg/tls13-spec/issues/1207) against the TLS 1.3 update to track this question.

> Section 4.2.1: This section defines, e.g., KE1 = credential_request, nonceU, info1*, epkU, so that KE1 contains the credential_request. However, at the top of section 4, the high-level protocol shows credential_request being sent in addition to KE1. This is another reason it would be helpful to have a complete diagram of the generic protocol (with, e.g., a generic 3-message AKE). For this particular issue, I think KE1 should only include AKE message components; it should not include the credential_request. Similarly for KE2 and section 4.3.1.

This is an interesting point. Mixing them together was done to highlight that the OPAQUE messages MUST be included in the AKE transcript. I'll see if we can show this requirement differently without losing readability.

> Section 6.6: What does “skipping the key exchange part” mean? Does it mean somehow truncating the protocol/skipping the AKE entirely? Or does it mean not using the exchanged key after a successful authentication? In general, I’m not sure what I would have to do to violate this section, which means I could easily violate it without knowing.

It means, for example, just running the OPRF flow and then doing nothing with the recovered credentials. Do you think it would be useful or valuable to be more specific here?

**Comment by crockeea** [comment]
- For the OPRF `KeyGen`, my confusion stemmed from looking in the OPRF API list and not finding `KeyGen` there. I propose adding the following to the OPRF API list:
   ```
   KeyGen(): Generates an OPRF private and public key. OPAQUE only requires an OPRF private key. We write (kU, _) = KeyGen() to denote use of this function for generating secret key kU (and discarding the corresponding public key).
   ```
   and removing the corresponding text from the paragraph below the list.
- For `Credentials`: I'm not very familiar with standard practices here, but should we perhaps point out that this struct is implementation-specific? Otherwise it just looks like it's missing/an error.
- "variant of CleartextCredentials": Section 3.1 contains two definitions of the `CleartextCredentials` struct, and `EnvelopeMode` is an enum that indicates which struct type is contained in the envelope (as I write this it strikes me that this is really not right: `CleartextCredentials` are not part of the envelope (at least not the `Envelope` struct)!). Can we get away with using a single `CleartextCredentials` struct? This eliminates the need for the `EnvelopeMode` enum, which results in a simpler and easier-to-understand protocol. I'm not seeing what benefit we get from defining two different `CleartextCredentials` structs with an enum to switch between them.
- On a related note: section 3.1 also defines `SecretCredentials` twice, but there are no differences between the two definitions. If you do end up keeping two structs for `CleartextCredentials` for some reason, then section 3.1 should define `SecretCredentials` only once. and view `EnvelopeMode` as switching between the two definitions of `CleartextCredentials`. (The current approach is to use the `EnvelopeMode` to switch between alternate definitions of both `*Credentials` types.)
- "Core Protocol diagram": To me, the "core" protocol is the combination of the OPRF and the AKE, and section 3 purports to describe the "Core Protocol". However, if you're proposing replacing the textual description with a diagram similar to section 3.3 augmented with the AKE messages, I'd be okay with that. As long as the complete (generic) diagram is somewhere.
- @hugokraw Regarding the paragraph about quantum safety: "OPAQUE may be instantiated....[h]owever, an instantiation where both AKE and data encryption are quantum safe, but the OPRF is not, would still ensure..." I read that as "an instantiation [of OPAQUE]", and when I think about the security of "OPAQUE", I think about the authentication/key exchange protocol itself, _not_ what happens after that (which seems to me to be beyond the scope of this specification). We should be clear which encryption scheme this is referring to.
- “skipping the key exchange part”: Yes, it would be valuable to be more specific here. I see two possible interpretations of the current paragraph. It either addresses a situation where
   - someone executes OPAQUE (by which I mean the OPRF+AKE), gets the key, and then throws the key away
   - someone executes only an OPRF (or perhaps what you currently refer to as the "core protocol") without any sort of key exchange

I think most of my confusion on several of these points arises because the document appears to define the "core protocol" as just the OPRF portion. I'm not sure why, since, as you point out, the AKE is _required_ for security. As written, the AKE comes off as a "second-class citizen". By including the AKE as part of the "core protocol", we make it clear that the AKE is a required part of OPAQUE. There is no longer the possibility of "skipping the key exchange", because it would be clear that a protocol that does an OPRF without an AKE is _not_ OPAQUE.

We should also more clearly define the scope of the specification. It refers only to the key exchange and authentication, and not to anything outside that (i.e., usage of OPAQUE in a larger protocol). This would, for example, clear up any confusion about which encryption scheme is referred to in the post-quantum security paragraph: it could only refer to the encryption scheme in the OPAQUE protocol itself.

**Comment by hugokraw** [comment]
> Regarding the paragraph about quantum safety: "OPAQUE may be instantiated....[h]owever, an instantiation where both AKE and data encryption are quantum safe, but the OPRF is not, would still ensure..." I read that as "an instantiation [of OPAQUE]", and when I think about the security of "OPAQUE", I think about the authentication/key exchange protocol itself, not what happens after that (which seems to me to be beyond the scope of this specification). We should be clear which encryption scheme this is referring to.

I suggest moving this text to the security considerations as subsection titled "Quantum-safe instantiation" with the following edited text:

OPAQUE may be instantiated with any post-quantum (PQ) AKE protocol with the security properties (KCI resistance and forward secrecy) outlined in {{security-considerations}}. To obtain a quantum-safe OPAQUE instantiation, also the OPRF needs to be quantum safe. Recent progress towards defining such OPRF can be found in [cite https://eprint.iacr.org/2020/1532]. We do stress that an OPAQUE instantiation where the AKE protocol is quantum-safe but the OPRF is not, would still ensure data security against future quantum attacks since breaking the OPRF does not retroactively affect the security of data transferred over a quantum-safe secure channel.

@crockeea @chris-wood Is that better? "quantum-safe secure channel." is short for "data transferred under a  quantum-safe encryption scheme with a key derived from a quantum-safe AKE". You may want to be more explicit.

**Comment by hugokraw** [comment]
> Section 6.6: What does “skipping the key exchange part” mean? Does it mean somehow truncating the protocol/skipping the AKE entirely? Or does it mean not using the exchanged key after a successful authentication? In general, I’m not sure what I would have to do to violate this section, which means I could easily violate it without knowing.

I suggest the following replacement text for that section:

OPAQUE provides PAKE (password-based authenticated key exchange) functionality in the client-server setting. While for user identification focus is often on the authentication part, we stress that the key exchange element is essential for data protection.  Authenticating the user but not protecting the data in the ensuing session is analogous to carefully checking a visitor's 
 credential at the door and then leaving the door open for others to enter freely.

@crockeea @chris-wood Is this better? I am ok with removing this remark all together. I heard this question a couple of times (why we need key exchange) so I wanted to make this point. But if it is confusing we can do without it.

**Comment by hugokraw** [comment]
From @crockeea 
> I think most of my confusion on several of these points arises because the document appears to define the "core protocol" as just the OPRF portion. I'm not sure why, since, as you point out, the AKE is required for security. As written, the AKE comes off as a "second-class citizen". By including the AKE as part of the "core protocol", we make it clear that the AKE is a required part of OPAQUE. 

@chris-wood @kevinlewi We must make sure that readers understand that there is no OPAQUE without AKE. So @crockeea suggestions in this regard are *very* important.

**Comment by bytemare** [comment]
+1

As an observer, I felt that the AKE part has been put aside, indeed. But I interpreted it as "we first need to thoroughly define secret credential recovery, and then we'll do the AKE parts".

As a reader and implementer, I have a lot of questions relative to the part post-secret-retrieval, as the draft doesn't go much into details, yet.

**Comment by crockeea** [comment]
> ) so I wanted to make this point. But if it is confusing we can d

I'm still confused by this one. I think we need to be more clear about the model to which you are referring. In particular, what does "data protection"/"protecting the data" refer to? It sounds like protection of data in a larger protocol, but that's not explicit, and it should be.

**Comment by hugokraw** [comment]
I was referring to the fact that most of the time the exchange between server and client does not stop at the success/failure of the user authentication but it is followed by some activity between client and server which typically carries some data. The KE part of PAKE is needed to derive a key that will protect such data. Given that this comment seems to generate more confusion than good, I propose removing it all together.

**Comment by bytemare** [comment]
"Data protection" may indeed be interpreted as regulatory compliance to "Data Protection" (e.g. GDPR and such).

Maybe replace this with

**Comment by bytemare** [comment]
"Data protection" may indeed be interpreted as regulatory compliance to "Data Protection" (e.g. GDPR and such).

Maybe replace this with

> OPAQUE provides PAKE (password-based authenticated key exchange) functionality in the client-server setting. While for user identification focus is often on the authentication part, we stress that the key exchange element is essential for providing means to encrypt subsequent the communication session. Authenticating the user but not providing confidentiality and integrity for the data in the ensuing session is analogous to carefully checking a visitor's credential at the door and then leaving the door open for others to enter freely.

**Comment by hugokraw** [comment]
I like this text. If others are happy with it, let's keep it. If it is still (potentially) confusing, I am ok with removing it all together.

**Comment by crockeea** [comment]
I'd like to keep something in the draft along these lines, but I am still confused. Specifically, med to be, or does the 
> we stress that the key exchange element is essential for providing means to encrypt subsequent the communication session.
Does this mean I have to protect subsequent data _with the OPAQUE key_, or is it just pointing out that confidentiality and integrity are important cryptographic properties? Could I obtain confidentiality and integrity with an independent key? For example, what if I did OPAQUE inside a TLS channel (not as part of the handshake)? Is this comment saying that I (still) _must_ use the OPAQUE key to protect data inside the TLS channel to obtain any guarantees about authentication?

**Comment by bytemare** [comment]
I see what you mean. Another aspect of this has been discussed [here](https://github.com/cfrg/draft-irtf-cfrg-opaque/issues/109#issuecomment-757061307).

How about this

> OPAQUE provides PAKE (password-based authenticated key exchange) functionality in the client-server setting. While for user identification focus is often on the authentication part, we stress that the key exchange element is essential for providing a shared secret. This shared secret is suitable for different use cases, e.g. session tokens or means to encrypt the subsequent communication session. Authenticating the user but not providing confidentiality and integrity for the data in the ensuing session is analogous to carefully checking a visitor's credential at the door and then leaving the door open for others to enter freely.

(Another paragraph would need to explain how to derive different keys for different usaged using HKDF-Expand.)

**Comment by hugokraw** [comment]
> Does this mean I have to protect subsequent data _with the OPAQUE key_, or is it just pointing out that confidentiality and integrity are important cryptographic properties? Could I obtain confidentiality and integrity with an independent key? For example, what if I did OPAQUE inside a TLS channel (not as part of the handshake)? Is this comment saying that I (still) _must_ use the OPAQUE key to protect data inside the TLS channel to obtain any guarantees about authentication?

This comment fully convinces me to drop this text. Eric is right about the possibility of running OPAQUE inside TLS without the need to generate a data-protection key by OPAQUE. Indeed, that is *exactly* how we plan to integrate OPAQUE with TLS. In this case, OPAQUE will serve as a "post-handshake authentication" of the TLS handshake without requiring to change the traffic keys established by the TLS handshake. It shows that in some cases it does make sense to run OPAQUE without deriving a data-protection key and therefore the text in question is, as Eric insisted, confusing.

**Comment by chris-wood** [comment]
@crockeea do you think a diagram in the authentication section is still needed? If not, I'll close this issue, since the remaining tasks were addressed.

**Comment by chris-wood** [comment]
I'm going to close this as resolved for now. @crockeea, if you think we still need to address points raised in this issue, can you please file separate issues for them? Thanks!
